packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,752 @@
+export interface Vis {
+  canvas?: HTMLCanvasElement;
+  colors: string[];
+  analyser?: AnalyserNode;
+  audioContext?: AudioContext; //butterchurn need it
+  oscStyle?: "dots" | "solid" | "lines";
+  bandwidth?: "wide" | "thin";
+  coloring?: "fire" | "line" | "normal";
+  peaks?: boolean;
+  safalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa_peak_falloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa?: "analyzer" | "oscilloscope" | "none";
+}
+import { out_spectraldata, renderHeight, renderWidth, windowShade, PIXEL_DENSITY, doubled } from "./Vis";
+
+let sapeaks = new Int16Array(76).fill(0);
+let sadata2 = new Float32Array(76).fill(0);
+let sadata = new Int16Array(76).fill(0);
+let safalloff = new Float32Array(76).fill(0);  
+let sample = new Float32Array(76).fill(0);
+let barPeak = new Int16Array(76).fill(0); // Needs to be specified as Int16 else the peaks don't behave as they should
+let i: number;
+let uVar12: number;
+let falloff: number; 
+let peakfalloff: number;
+
+let pushdown : number = 0;
+
+let logged: boolean = false;
+
+let colorssmall: string[] = [];
+let colorssmall2: string[] = [];
```

Comment:
I'm not sure about all these mutable globals. Feels like we're going to have bugs if there are multiple Webamps on the screen at the same time. Is it possible for these to be properties of the individual VisPaintHandler class instances as needed?

Reply:
Good call, I'll see how I can pull that off (it's only used for the spectrum analyzer anyway)